fly.io/cdn-2024-01-26/default.vcl

```diff
@@ -0,0 +1,117 @@
+# https://varnish-cache.org/docs/7.4/reference/vcl.html#versioning
+vcl 4.1;
+
+import std;
+
+# Thanks Matt Johnson! 👋
+# - https://github.com/magento/magento2/blob/03621bbcd75cbac4ffa8266a51aa2606980f4830/app/code/Magento/PageCache/etc/varnish6.vcl
+# - https://abhishekjakhotiya.medium.com/magento-internals-cache-purging-and-cache-tags-bf7772e60797
+
+backend default {
+  .host = "top1.nearest.of.changelog-2024-01-12.internal";
```

Comment:
You might want to play with DNS anycast by trying `top2.nearest`, especially if you have more than 3-4 `changelog.internal` origin servers.

Alternatively, you think it's worth experimenting with Varnish-level shielding? Cache nodes in low-traffic areas will have low cache-hit ratios.

I wonder if there would be a way to use Fly's DNS to specify that the cache servers in the same datacenters as the origin servers are shields. 🤔

Reply:
We started with `<appname>.internal` which resulted in Varnish errors since `.host` is not able to handle multiple values. FTR, this is what that returns:

```console
host changelog-2024-01-12.internal
changelog-2024-01-12.internal has IPv6 address fdaa:0:4556:a7b:2f7:4d32:8516:2
changelog-2024-01-12.internal has IPv6 address fdaa:0:4556:a7b:2f7:89b9:c9b8:2
```

`top2.nearest` will behave the same since we are running 2 instances in a single region:

```console
host top2.nearest.of.changelog-2024-01-12.internal
top2.nearest.of.changelog-2024-01-12.internal has IPv6 address fdaa:0:4556:a7b:2f7:4d32:8516:2
top2.nearest.of.changelog-2024-01-12.internal has IPv6 address fdaa:0:4556:a7b:2f7:89b9:c9b8:2
```

Do you think that we should use something other than `.host` in Varnish?

I really like your idea of having a shield Varnish node in the same region as the origins, and then pointing all other varnish nodes to that shield. Provided that we can figure out how to template the varnish config at instance boot time, we could do the following:

- if `$FLY_REGION == $PRIMARY_REGION` (both are environment variables available on every Fly.io machine), set the backend .hosts to the values from `changelog-2024-01-12.internal` (we would need to figure out how to handle more than 1 value)
- if `$FLY_REGION != $PRIMARY_REGION` set the backend `.hosts` to the Varnish shield region, i.e. `$PRIMARY_REGION.$FLY_APP_NAME.internal` (this assumes that we want to run more than one shield instance in the same region for redundancy purposes)

WDYT?

---
[Fly.io `.internal` DNS](https://fly.io/docs/networking/private-networking/#fly-io-internal-dns)

Reply:
> Do you think that we should use something other than .host in Varnish?

Based on my (possibly outdated) understanding, yes. `.host` will limit you to a single IP address. If you want to use a `CNAME` to support anycast, you'll want a `dynamic` backend: https://github.com/nigoroll/libvmod-dynamic

The alternative, I suppose, would be to resolve the anycast address at build, then create a `.host` for each IP it returns. But then you'd have to recompile the VCL regularly. If `dynamic` isn't an option, I suppose we could put a daemon on the containers that polls `top2.nearest.of` and compiles the VCL every minute, then tells Varnish to reload the config.